src/rpc.rs

```diff
@@ -12,80 +12,125 @@
  ********************************************************************************/
 use crate::UPClientZenoh;
 use async_trait::async_trait;
-use std::{string::ToString, time::Duration};
+use std::{string::ToString, sync::Arc, time::Duration};
 use tracing::error;
-use up_rust::{RpcClient, RpcClientResult, UAttributesError, UMessage, UMessageError, UUri};
+use up_rust::{
+    communication::{CallOptions, RpcClient, ServiceInvocationError, UPayload},
+    LocalUriProvider, UAttributes, UCode, UMessageType, UPayloadFormat, UPriority, UStatus, UUri,
+};
 use zenoh::prelude::r#async::*;
 
+pub struct ZenohRpcClient {
+    transport: Arc<UPClientZenoh>,
+    uri_provider: Arc<dyn LocalUriProvider>,
+}
+impl ZenohRpcClient {
+    /// Creates a new RPC client for the Zenoh transport.
+    ///
+    /// # Arguments
+    ///
+    /// * `transport` - The Zenoh uProtocol Transport Layer.
+    /// * `uri_provider` - The helper for creating URIs that represent local resources.
+    pub fn new(transport: Arc<UPClientZenoh>, uri_provider: Arc<dyn LocalUriProvider>) -> Self {
+        ZenohRpcClient {
+            transport,
+            uri_provider,
+        }
+    }
+}
+
 #[async_trait]
-impl RpcClient for UPClientZenoh {
-    // CY_TODO: Should remove method in the future
-    async fn invoke_method(&self, _method: UUri, request: UMessage) -> RpcClientResult {
-        // Get Zenoh key
-        let source = *request.attributes.source.0.clone().ok_or_else(|| {
-            let msg = "attributes.source should not be empty".to_string();
-            error!("{msg}");
-            UMessageError::PayloadError(msg)
-        })?;
-        let zenoh_key = if let Some(sink) = request.attributes.sink.0.clone() {
-            self.to_zenoh_key_string(&source, Some(&sink))
-        } else {
-            self.to_zenoh_key_string(&source, None)
+impl RpcClient for ZenohRpcClient {
```

Comment:
Would it suffice to use the `InMemoryRpcServer` from up-rust as our server? If not, why not?

Similarly, would the other default uP-L2 implementations from up-rust work? It's just the `RpcClient` that needs to be implemented differently for Zenoh?

Reply:
Yes, you're right. It suffices to use `InMemoryRpcServer`, or even the RpcServer in our example code.
In fact, `InMemoryRpcClient` still works in our case, but I just want to provide another efficient one for Zenoh users.

Reply:
I'm not sure if you think this fits, then, but would it make sense to show that we are able to use the uP-L2 APIs from up-rust and feed in the UPTransportZenoh and everything works in the tests?

So for example, in addition to the tests you already have,

* UPTransportZenoh{configured as an RPC server} <-> ZenohRpcClient{holding UPTransportZenoh}
* UPTransportZenoh{configured as an RPC server} <-> UPTransportZenoh{configured as an RPC client}

You could add these:

* UPTransportZenoh{configured as an RPC server} <-> InMemoryRpcClient{holding UPTransportZenoh}
* InMemoryRpcServer{holding UPTransportZenoh} <-> UPTransportZenoh{configured as an RPC client}
* InMemoryRpcServer{holding UPTransportZenoh} <-> InMemoryRpcClient{holding UPTransportZenoh}
* InMemoryRpcServer{holding UPTransportZenoh} <-> ZenohRpcClient{holding UPTransportZenoh}

I guess this can work out to a lot of combinations in the test code. Perfect use case or test_case? Or bad idea altogether? :sweat_smile: 

Was just thinking that it'd be good to ensure that UPTransportZenoh fits these roles in the automated testing.

Reply:
Hmm... I'm not sure whether these L2 APIs should be tested in the transport layer. In theory, we should only ensure the L1 API works here and don't need to care about how others use L1 API, which will be too many cases.
Furthermore, I'm wondering if it's worth to provide our own L2 RpcClient here. Perhaps users want more consistent API usage among different transport implementations.

Reply:
FMPOV you only need to make sure that you implement the Transport Layer API correctly. In the end, that's the whole purpose of having a layered approach ...
You _may_ want to provide a more efficient RpcClient implementation if it is worth the effort. In that case you should also verify that you have implemented the RpcClient API correctly (as we do in up-rust).

Reply:
Then I might want to take the RpcClient away... It will keep the layer cleaner

Reply:
As I think both @sophokles73 and I were saying, if you find having a more efficient approach worthwhile, let's keep it.

But yes, that would imply needing to add tests as he wrote.